src/input/l1/rpc_stream.rs

```diff
@@ -256,14 +285,80 @@ impl RpcStream {
     ///
     /// If multiple providers are given for either protocol or both, the system will rotate between
     /// them if and when one provider fails.
-    pub async fn new(options: L1ClientOptions) -> Result<Self> {
-        let builder = RpcStreamBuilder::new(options)?;
+    pub async fn new(
+        options: L1ClientOptions,
+        stake_table: Address,
+        reward_contract: Address,
+    ) -> Result<Self> {
+        let builder = RpcStreamBuilder::new(options, stake_table, reward_contract)?;
         Ok(builder.build().await)
     }
 
     /// Get the Espresso stake table genesis block.
-    pub async fn genesis(&self, _stake_table: Address) -> Result<(L1BlockId, Timestamp)> {
-        todo!()
+    pub async fn genesis(&self, stake_table: Address) -> Result<(L1BlockId, Timestamp)> {
+        let stake_table_contract = StakeTable::new(stake_table, &self.builder.provider);
+
+        // Get the block number when the contract was initialized
+        let initialized_at_block = stake_table_contract
+            .initializedAtBlock()
+            .call()
+            .await
+            .map_err(|err| {
+                Error::internal().context(format!("Failed to retrieve initialization block: {err}"))
+            })?
+            .to::<u64>();
+
+        // Fetch the finalized block to verify the initialized block is less than finalized
+        let finalized_block = self
+            .builder
+            .provider
+            .get_block(BlockId::finalized())
+            .await
+            .map_err(|err| {
+                Error::internal().context(format!("Failed to fetch finalized block: {err}"))
+            })?
+            .ok_or_else(|| Error::internal().context("Finalized block not found"))?;
+
+        let finalized_block_number = finalized_block.header.number;
+
+        if initialized_at_block >= finalized_block_number {
```

Comment:
I think if `initialized_at_block == finalized_block_number` it's fine? We should only panic if strictly `>`